src/native/s3select/s3select_napi.cpp

```diff
@@ -0,0 +1,251 @@
+/* Copyright (C) 2016 NooBaa */
+
+#include "../../../src/native/third_party/isa-l/include/crc.h"
+#include "../../../submodules/s3select/include/s3select.h"
+#include "../util/common.h"
+#include "../util/napi.h"
+#include <arpa/inet.h>
+
+namespace noobaa
+{
+
+enum InputFormat
+{
+    csv,
+    json,
+    parquet
+};
+
+class SelectWorker : public Napi::AsyncWorker
+{
+public:
+    Napi::Promise::Deferred deferred;
+    s3selectEngine::csv_object& csv_object;
+    s3selectEngine::json_object& json_object;
+    std::string select;
+    const char* buffer;
+    const unsigned int buffer_len;
+    const bool is_flush;
+    const InputFormat input_format;
+    const u_int8_t* headers_bytes;
+    const u_int32_t headers_len;
+
+    SelectWorker(Napi::Env env,
+        s3selectEngine::csv_object& csv_object,
+        s3selectEngine::json_object& json_object,
+        const char* buffer,
+        const unsigned int buffer_len,
+        const bool is_flush,
+        const InputFormat input_format,
+        const u_int8_t* headers_bytes,
+        const u_int32_t headers_len)
+        : AsyncWorker(env)
+        , csv_object(csv_object)
+        , json_object(json_object)
+        , buffer(buffer)
+        , buffer_len(buffer_len)
+        , is_flush(is_flush)
+        , input_format(input_format)
+        , deferred(env)
+        , headers_bytes(headers_bytes)
+        , headers_len(headers_len)
+    {}
```

Comment:
The worker has to hold a ref on the S3SelectNapi object to avoid gc which will make many pointers invalid here.
See how we did that in FSWrapWorker - https://github.com/noobaa/noobaa-core/blob/0a04de2f7e53289d42416d29f491940ce81ea8c7/src/native/fs/fs_napi.cpp#L488-L506

In addition, there is another lifetime issue about the args which needs to be handled like we do in FsWorker args_ref - https://github.com/noobaa/noobaa-core/blob/0a04de2f7e53289d42416d29f491940ce81ea8c7/src/native/fs/fs_napi.cpp#L378-L383

Lets discuss if this is unclear.